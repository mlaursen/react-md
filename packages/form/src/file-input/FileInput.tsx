import React, {
  forwardRef,
  Fragment,
  InputHTMLAttributes,
  ReactElement,
  ReactNode,
  Ref,
} from "react";
import { cnb } from "cnbuilder";
import { buttonThemeClassNames, ButtonThemeProps } from "@react-md/button";
import { TextIconSpacing, useIcon } from "@react-md/icon";
import { useInteractionStates } from "@react-md/states";
import { SrOnly } from "@react-md/typography";
import { bem } from "@react-md/utils";

type InputAttributes = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  "type" | "defaultValue" | "value"
>;

export interface FileInputProps extends ButtonThemeProps, InputAttributes {
  /**
   * An id for the input. This is required for a11y since it also is applied as
   * the `htmlFor` prop for the label.
   */
  id: string;

  /**
   * The change event handler to attach to this input. This is required since
   * there's really no use in this component otherwise.
   */
  onChange: React.ChangeEventHandler<HTMLInputElement>;

  /**
   * An optional icon to display for the file input.
   */
  icon?: ReactNode;

  /**
   * Boolean if the icon should appear after the children in the label.
   */
  iconAfter?: boolean;

  /**
   * Boolean if the children should not have some spacing between the icon and
   * itself.  The default behavior is to use the `<TextIconSpacing>` component
   * for text styled input buttons, but this can be disabled if you want to use
   * a screenreader only accessible label.
   */
  disableIconSpacing?: boolean;

  /**
   * Boolean if the file input should no longer allow the same file to be
   * selected multiple times and trigger the `onChange` each time it is
   * selected.
   */
  disableRepeatableFiles?: boolean;
}

const block = bem("rmd-file-input");

/**
 * This component is a wrapper for the `<input type="file" />` that can be themed
 * like a button.
 */
function FileInput(
  {
    style,
    className: propClassName,
    icon: propIcon,
    iconAfter = false,
    children = <SrOnly>Upload</SrOnly>,
    theme = "primary",
    themeType = "contained",
    buttonType = "icon",
    multiple = false,
    disableIconSpacing = false,
    disableRepeatableFiles = false,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseUp,
    onMouseLeave,
    onClick,
    onTouchStart,
    onTouchMove,
    onTouchEnd,
    onChange,
    ...props
  }: FileInputProps,
  ref?: Ref<HTMLInputElement>
): ReactElement {
  const { id, disabled } = props;
  const icon = useIcon("download", propIcon);

  const { ripples, className, handlers } = useInteractionStates({
    handlers: {
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onClick,
      onTouchStart,
      onTouchMove,
      onTouchEnd,
    },
    className: buttonThemeClassNames({
      theme,
      themeType,
      buttonType,
      disabled,
      className: propClassName,
    }),
    // pressing enter or space would trigger two click events otherwise.
    disableEnterClick: true,
  });

  let content: ReactNode = icon;
  if (disableIconSpacing || (children && !icon)) {
    content = (
      <Fragment>
        {!iconAfter && icon}
        {children}
        {iconAfter && icon}
      </Fragment>
    );
  } else if (children) {
    content = (
      <TextIconSpacing icon={icon} iconAfter={iconAfter}>
        {children}
      </TextIconSpacing>
    );
  }

  return (
    <Fragment>
      <input
        id={id}
        {...props}
        {...handlers}
        ref={ref}
        onChange={onChange}
        value={disableRepeatableFiles ? undefined : ""}
        type="file"
        className={block()}
        multiple={multiple}
      />
      <label
        htmlFor={id}
        style={style}
        className={cnb("rmd-file-input-label", className)}
      >
        {content}
        {ripples}
      </label>
    </Fragment>
  );
}

const ForwardedFileInput = forwardRef<HTMLInputElement, FileInputProps>(
  FileInput
);

if (process.env.NODE_ENV !== "production") {
  try {
    const PropTypes = require("prop-types");

    ForwardedFileInput.propTypes = {
      id: PropTypes.string.isRequired,
      name: PropTypes.string,
      onChange: PropTypes.func.isRequired,
      icon: PropTypes.node,
      iconAfter: PropTypes.bool,
      multiple: PropTypes.bool,
      disableIconSpacing: PropTypes.bool,
      disableRepeatableFiles: PropTypes.bool,
      theme: PropTypes.oneOf([
        "clear",
        "primary",
        "secondary",
        "warning",
        "error",
      ]),
      themeType: PropTypes.oneOf(["flat", "outline", "contained"]),
      buttonType: PropTypes.oneOf(["text", "icon"]),
      disabled: PropTypes.bool,
      children: PropTypes.node,
    };
  } catch (e) {}
}

export default ForwardedFileInput;
