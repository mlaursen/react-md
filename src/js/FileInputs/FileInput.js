import React, { PureComponent, PropTypes } from 'react';
import cn from 'classnames';

import injectInk from '../Inks';
import FontIcon from '../FontIcons';

/**
 * The `FileInput` component is used as simple styling for the `<input type="file" />`.
 * It will style the input as a raised button by default.
 */
class FileInput extends PureComponent {
  static propTypes = {
    /**
     * An optional style to apply.
     */
    style: PropTypes.object,

    /**
     * An optional className to apply.
     */
    className: PropTypes.string,

    /**
     * Boolean if the `FileInput` should be styled with the primary color.
     */
    primary: PropTypes.bool,

    /**
     * Boolean if the `FileInput` should be styled with the secondary color.
     */
    secondary: PropTypes.bool,

    /**
     * Boolean if the `FileInput` should be styled as a flat button instead of a
     * raised button.
     */
    flat: PropTypes.bool,

    /**
     * This should be a comma separated list of Media Types that the `FileInput` can
     * accept. If this prop is left blank, any file will be accepted.
     *
     * The values can either be:
     * - A file extension
     * - audio/*
     * - video/*
     * - image/*
     * - any valid [IANA Media Type](http://www.iana.org/assignments/media-types/media-types.xhtml)
     */
    accept: PropTypes.string,

    /**
     * Boolean if multiple files will be accepted.
     */
    multiple: PropTypes.bool,

    /**
     * A label to display on the `FileInput` when no files have been selected.
     */
    label: PropTypes.string.isRequired,

    /**
     * The icon children to use for the upload icon.
     */
    iconChildren: PropTypes.node,

    /**
     * The icon className to use for the upload icon.
     */
    iconClassName: PropTypes.string,

    /**
     * A function to call when the value of the input changes. This will
     * be triggered when the user selects a new file or cancels the new file selection.
     *
     * This function will be given the new [FileList](https://developer.mozilla.org/en-US/docs/Web/API/FileList)
     * as an array and the change event. If this is not a multiple file input, only the
     * newly selected File will be given instead of a list of one file. Since this is an
     * `input` tag, the user will not be able to select the same file multiple times unless
     * you manually clear the input's value.
     *
     * > NOTE: If the user hits cancel, null will be given for a single file input.
     *
     * ```js
     * onChange(files, e);
     * ```
     */
    onChange: PropTypes.func.isRequired,

    /**
     * An optional id for the file input field.
     */
    id: PropTypes.string,
  };

  static defaultProps = {
    label: 'Select a file from your computer',
    iconChildren: 'file_upload',
  };

  constructor(props) {
    super(props);

    this._handleChange = this._handleChange.bind(this);
  }

  _handleChange(e) {
    const { multiple, onChange } = this.props;
    const { files } = e.target;
    if (!multiple) {
      onChange(files[0] || null, e);
    } else {
      onChange(Array.prototype.slice.call(files), e);
    }
  }

  render() {
    const {
      style,
      className,
      label,
      iconChildren,
      iconClassName,
      primary,
      secondary,
      flat,
      id,
      ...props,
    } = this.props;
    delete props.onChange;

    return (
      <label
        style={style}
        className={cn(`md-btn md-${flat ? 'flat' : 'raised'}-btn md-file-input-btn`, className, {
          'md-primary': primary,
          'md-secondary': secondary,
        })}
        disabled={props.disabled}
        htmlFor={id}
      >
        <input
          {...props}
          id={id}
          aria-hidden="true"
          type="file"
          className="md-file-input"
          onChange={this._handleChange}
        />
        <div className="icon-separator">
          <FontIcon iconClassName={iconClassName} children={iconChildren} />
          <span className="text">{label}</span>
        </div>
      </label>
    );
  }
}

export default injectInk(FileInput);
