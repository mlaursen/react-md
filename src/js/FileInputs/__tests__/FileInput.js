/*eslint-env jest*/
jest.unmock('../FileInput');

import React from 'react';
import { findDOMNode } from 'react-dom';
import {
  Simulate,
  renderIntoDocument,
  findRenderedDOMComponentWithTag,
} from 'react-addons-test-utils';

import FileInput from '../FileInput';

const File = jest.fn(name => ({
  name: name || 'Test.jpg',
  lastModifiedDate: new Date(),
  isClosed: true,
  size: 39202,
  type: 'image/jpg',
}));

describe('FileInput', () => {
  it('merges className and style', () => {
    const style = { display: 'block' };
    const className = 'test';
    const fileInput = renderIntoDocument(
      <FileInput style={style} className={className} onChange={jest.fn()} />
    );

    const fileInputNode = findDOMNode(fileInput);
    expect(fileInputNode.style.display).toBe(style.display);
    expect(fileInputNode.className).toContain(className);
  });

  it('returns a single file when multiple is false onChange', () => {
    const onChange = jest.fn();
    const fileInput = renderIntoDocument(<FileInput onChange={onChange} />);

    const input = findRenderedDOMComponentWithTag(fileInput, 'input');

    // can't instantiate fully, so we can make an almost exact reference
    const files = [new File()];

    Simulate.change(input, { target: { files } });

    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toEqual(files[0]);
  });

  it('returns null when the user cancel\'s a file selection when multiple is false onChange', () => {
    const onChange = jest.fn();
    const fileInput = renderIntoDocument(<FileInput onChange={onChange} />);

    const input = findRenderedDOMComponentWithTag(fileInput, 'input');
    // Cancel returns empty list
    const files = [];

    Simulate.change(input, { target: { files } });

    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe(null);
  });

  it('returns a list of files when multiple is true onChange', () => {
    const onChange = jest.fn();
    const fileInput = renderIntoDocument(<FileInput onChange={onChange} multiple={true} />);

    const input = findRenderedDOMComponentWithTag(fileInput, 'input');

    // can't instantiate fully, so we can make an almost exact reference
    const files = [new File(), new File('Test2.jpg')];

    Simulate.change(input, { target: { files } });

    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toEqual(files);
  });

  it('returns an empty list of files when the user cancels an upload and multiple is true onChange', () => {
    const onChange = jest.fn();
    const fileInput = renderIntoDocument(<FileInput onChange={onChange} multiple={true} />);

    const input = findRenderedDOMComponentWithTag(fileInput, 'input');
    const files = [];

    Simulate.change(input, { target: { files } });

    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0].length).toBe(0);
    expect(onChange.mock.calls[0][0]).toEqual(files);
  });
});
