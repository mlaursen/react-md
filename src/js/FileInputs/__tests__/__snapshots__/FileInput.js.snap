// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput should apply styles correctly 1`] = `
<div
  class="md-inline-block md-file-input-container test-class"
  style="height: 200px;"
>
  <label
    aria-pressed="false"
    class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-btn md-btn--raised md-btn--text md-pointer--hover md-text"
    for="test-input"
    role="button"
    tabindex="0"
  >
    <div
      class="md-ink-container"
    />
    <div
      class="md-icon-separator"
    >
      <span
        class="md-icon-text"
      >
        Select a file
      </span>
      <i
        class="md-icon material-icons md-text--inherit"
      >
        file_upload
      </i>
    </div>
  </label>
  <input
    aria-hidden="true"
    class="md-file-input"
    id="test-input"
    tabindex="-1"
    type="file"
    value=""
  />
</div>
`;

exports[`FileInput should render the form control props correctly 1`] = `
<div
  class="md-inline-block md-file-input-container"
>
  <label
    aria-pressed="false"
    class="md-fake-btn md-pointer--hover md-fake-btn--no-outline md-btn md-btn--raised md-btn--text md-pointer--hover md-text"
    for="file-input-1"
    role="button"
    tabindex="0"
  >
    <div
      class="md-ink-container"
    />
    <div
      class="md-icon-separator"
    >
      <span
        class="md-icon-text"
      >
        Select a file
      </span>
      <i
        class="md-icon material-icons md-text--inherit"
      >
        file_upload
      </i>
    </div>
  </label>
  <input
    accept="images/*,videos/*"
    aria-hidden="true"
    class="md-file-input"
    id="file-input-1"
    name="images"
    tabindex="-1"
    type="file"
    value=""
  />
</div>
`;
